### PR TITLE
[CBRD-23868] Separate the routine of Java SP and Method

### PR DESCRIPTION
**Pull request by hgryoo**
http://jira.cubrid.org/browse/CBRD-23868

**Purpose**
As a preliminary work calling Java SP on the Server-Side, the routine to process Java SP should separate from the method.

**Implementation**
- [parser/parser_tree.h] `method_type` in `PT_METHOD_CALL` statement info struct
  * `class_or_inst` is changed to `method_type` which can have one of PT_IS_CLASS_MTHD, PT_IS_INST_MTHD, PT_SP_PROCEDURE and PT_SP_FUNCTION)
- [parser/name_resolution.c] `PT_METHOD_CALL` node is resolved with different value by the method type (javasp or C method) at `pt_resolve_stored_procedure()` and `pt_resolve_method()` respectively.
- [query/method_scan] METHOD_SCAN_ID (cubscan::method::scanner): method_scan is re-written in C++
   * METHOD_SCAN_BUFFER is removed
   * `method_sig_list` becomes packable_object and it is able to be packed by `cub_packing`
        * [query/method_def.cpp] pack(), unpack(), get_packed_size() are implmented  
   * Method arguments are sent as DB_VALUEs
        * VACOMM_BUFFER is removed
           * [network_interface_sr.c] `xs_send_action_to_client ()` is removed
        * VA_SCAN_ID is removed
   * The result of each method are received as DB_VALUE
- [query/query_method] Calling javasp or method routine in `method_invoke_for_server()`
   * method arguments, DB_VALUEs are stretched out with `method_sig_p->method_arg_pos[index]` for each method signature.
     The meaning _stretched out_ (I'm not sure exact term) : 
     for example) `SELECT test1(a, b), test2 (b, c) FROM tbl;`
      * Arguments are sent to CAS the form of (a, b, c) 
      * For test1, method_arg_pos will have (0, 1), For test2, method_arg_pos will have (1, 2)
      * So that the argument is stretched out by method signature as (a, b, c) -> (test1(a, b), test2(b,c))

**Remarks**
1) I found a bug at `pack_db_value()` in packer.cpp when packing SET type of db_value.
When calculating packed size, it calls `or_packed_value_size (&value, 1, 1, 0)`, which is `collapse_null:true`, `include_domain:true` and `include_domain_classoids:false`. 
However when DB_VALUE is packed by `pack_db_value()`, `or_pack_value()` is called with the following attributes; `collapse_null:false`, `include_domain:true` and `include_domain_classoids:true`.
It causes a buffer overflow, fixed at f83c490a3aa7fe3c3c492bffd83a6d9b0ea7809b

2) For stand-alone mode is not implemented yet in this PR. It will be done on another PR because the change list for code review is too long.

### CODE REVIEW

File: src/parser/name_resolution.c [hunk 1]
```diff
@@ -8933,19 +8947,16 @@ pt_resolve_method_type (PARSER_CONTEXT * parser, PT_NODE * node)
 
 
 /*
- * pt_make_method_call () - determines if the function call is really a
- *     method call and if so, creates a PT_METHOD_CALL to replace the node
- *     resolves the method call
+ * pt_make_method_call () - creates a PT_METHOD_CALL node and initilaizes from PT_FUNCTION node
  *   return:
  *   parser(in):
  *   node(in): an PT_FUNCTION node that may really be a method call
```
Comment (beyondykk9) [inline src/parser/name_resolution.c]: slip: 
node --> f_node

File: src/query/query_method.c [hunk 2]
```diff
@@ -24,510 +24,354 @@
...
- * method_initialize_vacomm_buffer () - Initializes the comm buffer
+ * method_send_error_to_server () - Send an error indication to the server
  *   return:
- *   vacomm_buffer(in)  :
- *   rc(in)     : client transmission request ID
- *   host(in)   :
- *   server_name(in)    :
+ *   rc(in)     : enquiry return code
+ *   host(in)   : host name
+ *   server_name(in)    : server name
```
Comment (beyondykk9) [inline src/query/query_method.c]: slip:
host --> host_p
server_name --> error_id

File: src/query/method_def.cpp [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    }
+  return size;
+}
+
+void
+method_sig_list::unpack (cubpacking::unpacker &deserializator)
+{
+  // assuming that METHOD_SIG_LIST is already allocated
+
+  deserializator.unpack_int (num_methods);
```
Comment (beyondykk9) [inline src/query/method_def.cpp:104]: not needed to check if num_methods is 0 or negative ?

File: src/query/method_def.cpp [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+	  if (i != num_methods - 1) /* last */
+	    {
+	      method_sig_p->next = (METHOD_SIG *) db_private_alloc (NULL, sizeof (METHOD_SIG));
+	      method_sig_p->next->next = nullptr;
+	    }
+	  method_sig_p = method_sig_p->next;
+	}
+    }
+}
```
Comment (beyondykk9) [inline src/query/method_def.cpp]: add a new line at the end of file.

File: src/query/query_method.c [hunk 5]
```diff
@@ -24,510 +24,354 @@
...
- * method_send_eof_to_server () -
- *   return:
- *   vacomm_buffer(in)  : Transmission buffer
- *
- * Note: Send the transmission buffer to the server and indicate EOF.
+ * method_invoke () -
+ *   return: int
+ *   result (out)     :
+ *   arg_vals (in)   : objects & arguments DB_VALUEs
+ *   method_sig_list (in) : Method signatures
```
Comment (beyondykk9) [inline src/query/query_method.c]: slip:
method_sig_list --> meth_sig_p

File: src/query/query_method.c [hunk 6]
```diff
@@ -24,510 +24,354 @@
...
 
-  for (count = 0, value_p = val_list_p; count < value_count; count++, value_p++)
-    {
-      db_make_null (value_p);
-    }
+/*
+ * method_fixup_set_vobjs() - if val is a set/seq of vobjs then
+ * 			    turn it into a set/seq of vmops
+ *   return: NO_ERROR on all ok, ER status( or ER_FAILED) otherwise
+ *   val(in/out): a db_value
```
Comment (beyondykk9) [inline src/query/query_method.c]: slip:
val --> value_p

File: src/query/query_method.c [hunk 7]
```diff
@@ -24,510 +24,354 @@
...
 
 /*
- * method_sig_list_freemem () -
- *   return:
- *   meth_sig_list(in)        : pointer to a meth_sig_list
- *
- * Note: Free function for METHOD_SIG_LIST using free_and_init.
+ * method_fixup_vobjs () -
+ *   return: NO_ERROR on all ok, ER status( or ER_FAILED) otherwise
+ *   value(in/out): a db_value
```
Comment (beyondykk9) [inline src/query/query_method.c]: slip:
value --> value_p

File: src/query/method_def.cpp (hunk 4 above)
Comment (shparkcubrid) [inline src/query/method_def.cpp]: no newline at end of file

File: src/query/method_scan.cpp [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+	    }
+	}
+
+      if (scan_code == S_SUCCESS)
+	{
+	  qproc_db_value_list *dbval_list = m_dbval_list;
+	  int num_methods = m_method_sig_list->num_methods;
+	  for (int i = 0; i < num_methods; i++)
+	    {
+	      DB_VALUE *dbval_p = (DB_VALUE *) malloc (sizeof (DB_VALUE));
```
Comment (shparkcubrid) [inline src/query/method_scan.cpp]: How about using 'db_private_alloc ()'?

File: src/query/method_def.cpp (hunk 3 above)
Comment (hgryoo) [reply src/query/method_def.cpp:104]: it is an unpacking method, `num_methods` value is set from the buffer

File: src/query/method_def.cpp (hunk 4 above)
Comment (hgryoo) [reply src/query/method_def.cpp]: Thank you

File: src/parser/name_resolution.c (hunk 1 above)
Comment (hgryoo) [reply src/parser/name_resolution.c]: fixed

File: src/query/query_method.c (hunk 2 above)
Comment (hgryoo) [reply src/query/query_method.c]: fixed

File: src/query/query_method.c (hunk 6 above)
Comment (hgryoo) [reply src/query/query_method.c]: fixed

File: src/query/query_method.c (hunk 5 above)
Comment (hgryoo) [reply src/query/query_method.c]: fixed

File: src/query/query_method.c (hunk 7 above)
Comment (hgryoo) [reply src/query/query_method.c]: fixed

File: src/query/method_scan.cpp (hunk 8 above)
Comment (hgryoo) [reply src/query/method_scan.cpp]: I patched at 87e91e0, thank you

**Review by shparkcubrid — APPROVED** [review]

**Review by ctshim — APPROVED** [review]

**Review by beyondykk9 — APPROVED** [review]

**Comment by hgryoo** [comment]
69c7379, I think the legacy code allocates the method arguments with `malloc()` so that the resource tracker couldn't check the memory leak. I followed the same routine but after using the private allocator(`db_private_alloc()`), The memory leak is check. so I fixed it.

**Review by youngjinj — APPROVED** [review]

**Pull request merged**